### ISSUE TITLE
System.out (and System.err) messages from a Process spawned from a build task are completely lost in Gradle daemon mode

### ISSUE THREAD

**Issue by jaikiran** [gradle/gradle#13522]
(This is going to be a bit long issue description, so please bear with me)

### Background:

Gradle builds run in a no-daemon as well as a daemon mode. This bug affects the daemon mode. When a gradle build is launched, in daemon mode, the `GradleMain` class internally triggers and launches a process with `GradleDaemon` as the entry point class. Effectively during the build process, we now have 2 processes - one the `GradleMain` which acts as a client and the other the `GradleDaemon` process. During the initial "handshake" phase between these 2 processes, the client gradle process consumes messages written out to the `STDOUT` of the daemon process and waits for a specific message to appear in that stream[1]. Once that has happened, the client process considers the handshake successful and moves on to doing other build related things. While doing so, it "abandons" listening for any content written out to the daemon process' `STDOUT`. I'll explain why this is a bad thing, a few sections later. Once this handshake is complete, these processes then communicate with each other over a socket connection. Any messages logged on the daemon process are passed along over this socket as a `LogEvent` which then gets processed  by this client process and gets printed out on this client process' build console (of course after the usual log level checks etc...). Additionally, this client process during its bootstrap process has also replaced the `System.out` with its internal logging system classes so that any `System.out.print...` messages, within that process are routed over its logging system which then goes on to log those messages. This part is fine - I mention it solely for the sake of completeness.

Coming to the "daemon" process, when it launches, it does a bunch of things. One of them being setting up its logging system. In doing so, it "switches"[1] the `System.out` and `System.err` streams in favour of an internal logging classes which pass along (and log) the messages that are issued using `System.out` and `System.err` during the lifetime of that process by the code running in that process instance. All is fine so far. Once it has switched the `System.out` and `System.err` it completes the handshake with the client process and then additionally goes on to close the original `System.out` and `System.err` streams[3]. This in itself shouldn't really be a big issue since anything logged via `System.out` in that process is already being redirected to a different (log based) stream. However, closing these streams can and does cause issues (which I explain below). But first let's consider a simple build process and see how it works.

### Normal builds:

Builds that are initiated by the gradle client process (remember the process which initially started this daemon) are run within this daemon process and the build tasks' code runs within this process. Any `System.out` messages issued within these tasks gets redirected to the internal gradle logging system which then passed it along over the socket to the client process, which as explained above, processes it and prints it if necessary.

### Builds that run into issue:

Like I noted, this handling of `STDOUT`/`System.out` etc... both in the client and daemon process doesn't show up as a problem, until there's a build task which internally spawns a `Process` and that then writes out to `System.out`/`System.err`. One such build task is the `native-image` build task belonging to the Quarkus project. In the context of this issue, it doesn't matter what exactly the implementation of this task looks like. The only important bit is that this task implementation[4] (belonging to the Quarkus project) ends up spawning a Java `Process` which internally issues messages to the `System.out`. Essentially, at this point, we now have 3 processes in this whole build - the first two are the gradle client and the gradle daemon as usual and now this 3rd one spawned from within a build task code running within the daemon. Here's where the problem starts. Any `System.out` issued by this 3rd processes (running as a child of the daemon) is completely lost. Crucial log messages issued during these build tasks are potentially lost[5]. The reason for this loss of messages, is because the chain of `STDOUT`/`STDERR` between this child process and the daemon and then the daemon and the gradle client process has been broken at multiple places. One, between the child process and the gradle daemon because of closing of the original `System.out` and `System.err` streams at [3]. These original underlying streams should be left open for cases like these where they get used by processes spawned from within the build tasks, so that the IO propagation (supported by Java) continues to work between (sub)processes. Fixing this part of the code (to not close these original streams) is fixing one part of the broken chain. The next place where a fix is needed is the code in [1] where the gradle client process shouldn't abandon the underlying `STDOUT` of the process. Instead it should continue to consume messages off it (in a thread) even after the handshake is complete, so that any messages that arrive on it can then be printed out by the client process into the build console. Once both these places are fixed, the leaf process (the one spawned by a build task) that intitiates any `System.out` messages will then be propagated to the gradle daemon's `STDOUT` which will then directly propagate to the `STDOUT` of the gradle client process, which then will/should be able to just log them using `System.out.println(...)` calls so that those messages then gets routed over its logging system (remember the gradle client process has replaced `System.out` with its internal logging system stream) and finally gets printed to the build console.

I have a patch which contains these fixes/changes and my local testing of this patch against Quarkus application builds has now shown that these messages do get logged/shown correctly and the Quarkus application build with Gradle is much more informative (and similar to Maven) now. Before I send out those patches, I will try and see how it behaves when the same `GradleDaemon` process is used by multiple client builds, either at the same time (is that possible?) or one after another. The successful tests, with this patch, that I ran so far have been with a "single use" gradle daemon where the daemon process stops itself after the client build is complete.

[1] https://github.com/gradle/gradle/blob/master/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/bootstrap/DaemonOutputConsumer.java#L58
[2] https://github.com/gradle/gradle/blob/master/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/bootstrap/DaemonMain.java#L195
[3] https://github.com/gradle/gradle/blob/master/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/bootstrap/DaemonMain.java#L149
[4] https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java#L339
[5] https://github.com/quarkusio/quarkus/issues/9306 (this one is just one example, I remember similar issues being reported, when Gradle was used to build Quarkus applications)

**Comment by big-guy** [comment]
@jaikiran Thanks for the details.

> These original underlying streams should be left open for cases like these where they get used by processes spawned from within the build tasks, so that the IO propagation (supported by Java) continues to work between (sub)processes. Fixing this part of the code (to not close these original streams) is fixing one part of the broken chain. The next place where a fix is needed is the code in [1] where the gradle client process shouldn't abandon the underlying STDOUT of the process. Instead it should continue to consume messages off it (in a thread) even after the handshake is complete, so that any messages that arrive on it can then be printed out by the client process into the build console. Once both these places are fixed, the leaf process (the one spawned by a build task) that intitiates any System.out messages will then be propagated to the gradle daemon's STDOUT which will then directly propagate to the STDOUT of the gradle client process, which then will/should be able to just log them using System.out.println(...) calls so that those messages then gets routed over its logging system (remember the gradle client process has replaced System.out with its internal logging system stream) and finally gets printed to the build console.

I think this won't work for a few reasons:
- I think this could cause Gradle to duplicate anything logged with `System.out` or `System.err` in the build process
- Gradle used to do something similar to this, but it caused other problems with output from one task getting mixed into the output of another task because you would now have multiple things trying to write to the same console.  
- Gradle tries to associate output from a task with that task so that it gets grouped properly in the console and shows up in build scans under the right task.  I think this would cause that grouping to break. 
- Both of these problems get worse with multiple NativeImageBuildSteps happening at once.

Is it possible for Quarkus NativeImageBuildStep to either use Gradle's APIs to run the process (exec or Worker API) or for NativeImageBuildStep to expose stdout/stderr as parameters? That would let something that uses NativeImageBuildStep to repipe the output to something that would be captured by Gradle.

**Comment by jaikiran** [comment]
Hello @big-guy, thank you for the useful inputs.

> Is it possible for Quarkus NativeImageBuildStep to either use Gradle's APIs to run the process (exec or Worker API) or for NativeImageBuildStep to expose stdout/stderr as parameters? That would let something that uses NativeImageBuildStep to repipe the output to something that would be captured by Gradle.

That piece of code is in a layer where, AFAIK, we avoid any build tool specific code (both Maven and Gradle build tasks ultimately call this build tool agnostic code). So adding some kind of logic which requires Gradle specific APIs (like the `Project#exec` or the worker APIs) wouldn't be feasible. Furthermore, I did look at the `exec` API available on the gradle `Project` and although it probably will help for cases like these, I think the returned `ExecHandle` is far too minimal and hides some of the useful APIs that `java.lang.Process` provides (starting Java 9). Having `exec` return the underlying `Process` would be more useful IMO. The other limitation of that API, IMO, is that it's a blocking call and waits for the process to complete, so it isn't suited in many cases.

Given all this and the fact that I probably won't be able to use these Gradle specific code in that part of the code, I came up with a workaround and have submitted a PR for that in Quarkus https://github.com/quarkusio/quarkus/pull/10137. All it does is, streams the `STDOUT`/`STDERR` over the `System.out`/`System.err` (which Gradle has already replaced by its logging system), once that part of the code gets a hint that it's running as part of the Gradle build task. It has got us past that issue.

Like I note in that PR, I was initially planning to submit a PR here which would have stopped closing the underlying streams and also would have updated the `DaemonOutputConsumer` to continue consuming after the handshake, but given your input:
> I think this won't work for a few reasons:
>  -   I think this could cause Gradle to duplicate anything logged with System.out or System.err in the build process
>   -  Gradle used to do something similar to this, but it caused other problems with output from one task getting mixed into the output of another task because you would now have multiple things trying to write to the same console.
>   -  Gradle tries to associate output from a task with that task so that it gets grouped properly in the console and shows up in build scans under the right task. I think this would cause that grouping to break.

I've shelved that plan, since I don't think my proposed PR would have managed to solve these above issues.

**Comment by ljacomet** [comment]
There is no current plan to solve this in Gradle itself. As it looks like a solution was found for Quarkus, closing here.
If there are further use cases that require this to be implemented on the Gradle side, please comment back and we will consider opening it up again.
